packages/php-wasm/universal/src/lib/journal-memfs.ts

```diff
@@ -0,0 +1,266 @@
+	};
+
+	const originalTruncate = FS.truncate;
+	FS.truncate = function (path: string, ...rest: any[]) {
+		let node;
+		if (typeof path == 'string') {
+			const lookup = FS.lookupPath(path, {
+				follow: true,
+			});
+			node = lookup.node;
+		} else {
+			node = path;
+		}
+		addEntry({
+			operation: 'UPDATE_FILE',
+			path: FS.getPath(node),
+		});
+		return originalTruncate(path, ...rest);
+	};
+
+	const originalUnlink = FS.unlink;
+	FS.unlink = function (path: string, ...rest: any[]) {
+		addEntry({
+			operation: 'DELETE',
+			path,
+			nodeType: 'file',
+		});
+		return originalUnlink(path, ...rest);
+	};
+
+	const originalMkdir = FS.mkdir;
+	FS.mkdir = function (path: string, ...rest: any[]) {
+		addEntry({
+			operation: 'CREATE_DIRECTORY',
+			path,
+		});
+		return originalMkdir(path, ...rest);
+	};
+
+	const originalRmdir = FS.rmdir;
+	FS.rmdir = function (path: string, ...rest: any[]) {
+		addEntry({
+			operation: 'DELETE',
+			path,
+			nodeType: 'directory',
+		});
+		return originalRmdir(path, ...rest);
+	};
+
+	const workingSet: Set<string> = new Set();
+	function addEntry(entry: FilesystemOperation) {
+		if (!entry.path.startsWith(memfsRoot) || entry.path === memfsRoot) {
+			return;
+		}
+		// Partition the entries into groups of operations that can be
+		// performed in parallel.
+		const lastPartition = journal.partitions[journal.partitions.length - 1];
+		const lastEntry = lastPartition[lastPartition.length - 1];
+
+		// RENAME entries are always added to a new partition because
+		// parallelizing the following two RENAME operations only works
+		// if the first one is done before the second one:
+		//     RENAME /path_1 /path_2
+		//     RENAME /path_2 /path_3
+		if (
+			entry.operation === lastEntry.operation &&
+			entry.operation !== 'RENAME'
+		) {
+			// Don't add duplicate entries to the same partition.
+			// to make async processing easier later on.
+			if (workingSet.has(entry.path)) {
+				return;
+			}
+			workingSet.add(entry.path);
+			lastPartition.push(entry);
+			onEntry(entry);
+			return;
+		}
+
+		journal.partitions.push([entry]);
+		workingSet.clear();
+		onEntry(entry);
+	}
+
+	const journal = {
+		partitions: [
+			[{ operation: 'NOOP', path: '' }],
+		] as FilesystemOperation[][],
+		size() {
+			return journal.partitions
+				.map((p) => p.length)
+				.reduce((a, b) => a + b, 0);
+		},
+		flush(): FilesystemOperation[][] {
+			// Remove the NOOP partition
+			const entries = this.partitions.slice(1);
+			journal.partitions = [[{ operation: 'NOOP', path: '' }]];
+			return entries;
+		},
+		unbind() {
+			FS.write = originalWrite;
+			MEMFS.ops_table.dir.node.rename = originalRename;
+			FS.truncate = originalTruncate;
+			FS.unlink = originalUnlink;
+			FS.mkdir = originalMkdir;
+			FS.rmdir = originalRmdir;
+			FS.hasJournal = false;
+		},
+	};
+	return journal;
+}
+
+function normalizeMemfsPath(path: string) {
+	return path.replace(/\/$/, '').replace(/\/\/+/g, '/');
+}
```

Comment:
do you think it would make sense to try out a [Proxy object](http://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy) here instead of re-implementing the existing FS?

I like this Journal-MEMFS filesytem, but a proxy could be interesting too.

Reply:
Oh I like that idea, it could work

Reply:
Actually, Emscripten JavaScript module uses a hard-coded references to the FS object so we can't just replace it with a proxy object here. Nonetheless these explorations resulted in a cleaner implementation which is nice.